packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -0,0 +1,154 @@
+ * @alpha
+ */
+export interface AbortTransaction {
+	/**
+	 * The failed outcome indicating the rest of the transaction should abort and any changes should be rolled back.
+	 * This will only fail the current transaction scope. Any transactions in the outer scope will still run.
+	 */
+	readonly result: "abort";
+}
+
+/**
+ * The extended successful outcome of a transaction, i.e. the transaction should continue.
+ * @alpha
+ */
+export interface ContinueTransactionExt<TSuccessValue> extends ContinueTransaction {
+	/** The user defined return value on successfully completing the transaction */
+	readonly returnValue: TSuccessValue;
+}
+/**
+ * The extended failed outcome of a transaction, i.e. the transaction should abort.
+ * @alpha
+ */
+export interface AbortTransactionExt<TFailureValue> extends AbortTransaction {
+	/** The user defined return value on failing the transaction */
+	readonly returnValue: TFailureValue;
+}
+
+/**
+ * The successful return value of the runTransaction API, i.e., the transaction succeeded.
+ * @alpha
+ */
+export interface RunTransactionSucceeded {
+	/** Property indicating that the transaction was successful */
+	readonly success: true;
+}
+/**
+ * The failed return value of the runTransaction API, i.e., the transaction failed.
+ * @alpha
+ */
+export interface RunTransactionFailed {
+	/** Property indicating that the transaction failed */
+	readonly success: false;
+}
+
+/**
+ * The extended successful return value of the runTransaction API, i.e., the transaction succeeded.
+ * @alpha
+ */
+export interface RunTransactionSucceededExt<TSuccessValue> extends RunTransactionSucceeded {
+	/** The user defined return value on successfully completing the transaction */
+	readonly returnValue: TSuccessValue;
+}
+/**
+ * The extended failed return value of the runTransaction API, i.e., the transaction failed.
+ * @alpha
+ */
+export interface RunTransactionFailedExt<TFailureValue> extends RunTransactionFailed {
+	/** The user defined return value on failing the transaction */
+	readonly returnValue: TFailureValue;
+}
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface RunTransactionParams {
+	/**
+	 * The function to run as the body of the transaction.
+	 * @returns The result of the transaction indicating whether the transaction should continue or abort.
+	 */
+	readonly transaction: () => ContinueTransaction | AbortTransaction;
+	/**
+	 * An optional list of {@link TransactionConstraint | constraints} that are checked just before the transaction begins.
+	 * If any of the constraints are not met when `runTransaction` is called, it will throw an error.
+	 * If any of the constraints are not met after the transaction has been ordered by the service, it will be rolled back on
+	 * this client and ignored by all other clients.
+	 */
+	readonly preconditions?: readonly TransactionConstraint[];
```

Comment:
The fact that we have two variants of `RunTransactionParams` (maybe three, see comment above) makes me want to put `preconditions` in on a `BaseRunTransactionParams` interface (final name TBD) that all the `RunTransactionParams` extend.

The alternative would be to extract the `transaction` parameter into its own parameter. This would lead to: `runTransaction(transaction: () => ContinueTransaction | AbortTransaction, params?: RunTransactionParams): RunTransactionSucceeded | RunTransactionFailed`
That would reduce bring the number of `RunTransactionParams` types back down to one. I think this is a pretty common pattern: use mandatory arguments for the things that must be provided, and use a single object for the optional config stuff.